### DISCUSSION TITLE
What considerations to handle a post method for a contact form for Nextjs??

### DISCUSSION THREAD

**Discussion by Elindo586** [vercel/next.js#37808] [Help] [answered]
So I haven't done any of the backend with Next.js..

What would be different on handling a post method for a contact form using React vs. Next.js? 

Usually on React I might need Express, Cors, dotenv... maybe something else?  to handle a post method and send the form to my email..

Would Next.js work the same?... or are there any differences to keep in mind when coding the backend for this post request to submit a contact form? 

Would there happen to be any code samples for this???

Also... in a tangent topic.....  what would be a good way to stop spam from a contact form?  I am thinking just manually add emails to be blocked somewhere...

**Comment by icyJoseph** [comment]
Next.js is a React framework, I think when you say React you mean, CRA, or Vite? in CSR mode? Not very sure about the language there.

All you need to do is create `/api/contact` and submit the form to that endpoint, either natively or with fetch. For example (simplified use case):

```jsx
const Home = () => {
  return (
    <div>
      <form
        onSubmit={async (e) => {
          e.preventDefault();

          const username = e.currentTarget?.["username"].value || "";

          await fetch("/api/contact", {
            method: "POST",
            body: JSON.stringify({ username }),
            headers: { "Content-Type": "application/json" },
          });
        }}
      >
        {/* form fields */}
        <input type="text" name="username" />

        <button type="submit">Submit</button>
      </form>
    </div>
  );
};

export default Home;
```

And then in `api/contact`

```ts
export default function handler(req, res) {
  if (req.method !== "POST") {
    return res.status(400).end();
  }
  console.log(req.body);
  // do whatever you need with the data
  return res.status(200).json({ ...req.body });
}
```

Working with the FormData type needs a bit more work, but it is not different than what you'd do with other frameworks. It's just JavaScript. I'd say go with a JSON payload to get started.

I'd also recommend looking at https://react-hook-form.com/.


> There is also this package, https://next-runtime.meijer.ws/getting-started/1-introduction , **but I think it is best to learn to do this without help from packages**. The aim of this package is to avoid having to create yet another `api/contact` and use the page itself.


Last but not least, keep asking questions as you develop this. Glad to help.

**Comment by Elindo586** [comment]
Thanks much!

Say...  when I say  React I am talking about a typical CRA  (create-react-app)...  when I use CRA and specially if I am working on the backend I also have to install Express, Cors, dotenv, and nodemon...   are any of these plugins included in create-next ?  


So far I have only done front end work with Next, and the only plugins I have installed are bootstrap and gsap for animation.

**Comment by icyJoseph** [comment, marked as answer]
Next.js is already a server, with a very nice feature called, API routes. Abundant documentation can be found here, https://nextjs.org/docs/api-routes/introduction.

These API routes are like express routes, where the route is given by the filename system. These all live inside the `pages/api` folder. All methods arrive to them, so it is usually good to limit that. 

They run on demand, and have access to the `.env` vars your load for your application. Since they live in the same domain as the server, CORS is not needed, unless special cases. They are controlled by the Next.js server. So no, express, no cors, no dotenv, no nodemon are needed, what so ever. 

Give it a go yourself, create a `pages/api/hello.js` file:

```js
export default function handler(req, res) {
  res.status(200).json({ hello: 'World' })
}
```

And from the browser, open the console and just write `fetch("/api/hello").then(res=>res.json()).then(console.log)`, it works. 

Inside this `pages/api` routes, you can work in 100% Node.js environment, so you can connect to databases, email streams, etc... they do have a size limit, but you'll hardly reach it.

**Comment by Elindo586** [comment]
Thanks...

Say... when you say limit, what limits are we talking about?  I am planning to use a few more json files... later I will just put them into API's..


Also.,... pretty good information on the React hook form.... where / how do you find all those resources?  speaking of... would you have information like that for pagination and input search boxes? .......     I want to implement those features later on for blogs..

Oh and yes...  I try to use a little plugins as possible  because I don't want to be dealing with upgrading old stuff in the future.. or skipping  on some some learning...

**Reply by icyJoseph** [reply]
Limits of running on Vercel for example. https://nextjs.org/docs/messages/api-routes-response-size-limit

Other than that, you are cool. Also, remember that this `api/routes` are lambdas/tasks/serverless, so you should never assume that the same lambda is serving your request. The server creates them and kills them as needed.

> would you have information like that for pagination and input search boxes? 

You mean like, as you type searching with pagination? You could use [react-query](https://github.com/TanStack/query) or [swr](https://github.com/vercel/swr) - not really sure what you meant there though, but these two can handle querying for paginated data indexed by a key (the query).
